source/GameInterface/Services/Kingdoms/Handler/KingdomUpdateHandler.cs

```diff
@@ -0,0 +1,68 @@
+﻿using Common.Logging;
+using Common.Messaging;
+using Common.Network;
+using GameInterface.Services.Kingdoms.Messages;
+using GameInterface.Services.Kingdoms.Patches;
+using GameInterface.Services.ObjectManager;
+using Serilog;
+using TaleWorlds.CampaignSystem;
+using static TaleWorlds.CampaignSystem.Actions.ChangeKingdomAction;
+
+namespace GameInterface.Services.Kingdoms.Handler
+{
+    /// <summary>
+    /// Handles all changes to kingdoms.
+    /// </summary>
+    public class KingdomUpdateHandler : IHandler
+    {
+        private readonly IMessageBroker messageBroker;
+        private readonly ILogger Logger = LogManager.GetLogger<KingdomUpdateHandler>();
+
+        public KingdomUpdateHandler(IMessageBroker messageBroker, INetwork network, IObjectManager objectManager)
+        {
+            this.messageBroker = messageBroker;
+            messageBroker.Subscribe<UpdatedKingdomRelation>(Handle);
+            messageBroker.Subscribe<PolicyAdded>(Handle);
+            messageBroker.Subscribe<PolicyRemoved>(Handle);
+        }
+
+        public void Dispose()
+        {
+            messageBroker.Unsubscribe<UpdatedKingdomRelation>(Handle);
+            messageBroker.Unsubscribe<PolicyAdded>(Handle);
+            messageBroker.Unsubscribe<PolicyRemoved>(Handle);
+        }
+
+        private void Handle(MessagePayload<UpdatedKingdomRelation> obj)
+        {
+            var payload = obj.What;
+
+            Clan clan = Clan.FindFirst(x => x.StringId == payload.ClanId);
+            Kingdom kingdom = Kingdom.All.Find(x => x.StringId == payload.KingdomId);
```

Comment:
This is super expensive, the kingdom should exist in the IObjectManager let me know if not